rfcs/2018-11-23-annotations.md

```diff
@@ -0,0 +1,406 @@
+    /// those annotations which intersect the given interval.
+    fn iter_range(&self, interval: Interval) -> impl Iterator<Item=AnnotationSlice> { }
+    /// Removes any annotations provided by this plugin
+    fn clear(&mut self, plugin: PluginId) { }
+}
+```
+
+When updating the frontend, we will take the current visible range, and
+grab all the annotations that intersect that range:
+
+```rust
+// in the function preparing the update (say send_update_for_plan)
+// let ops = something;
+// let pristine = sure;
+
+let start_off = self.offset_of_line(self.first_line);
+let end_off = self.offset_of_line(self.first_line + self.height + 1);
+let visible_range = Interval::new(start_off, end_off);
+let annotations = iter::once(self.selection.to_annotation(visible_range)
+    .chain(self.find.iter().map(|f| f.to_annotation(visible_range))
+    .chain(self.annotations.iter_range(visible_range)
+    .collect::<Vec<_>>();
+
+let params = json!({
+        "ops": ops,
+        "pristine": pristine,
+        "annotations": annotations,
+    });
+
+client.update_view(self.view_id, &params);
+```
+
+***future work***: At some point we may want to explore something like
+minimal invalidation for annotations; for a simple approach we could
+imagine not resending annotations of a given type when they are
+unchanged. For the time being, the simplicity of being entirely
+stateless is appealing.
+
+
+### In the frontend
+
+The annotation wire format is intended to be handled efficiently by
+frontends, particularly when drawing. Specifically, the annotations sent
+should include only those necessary to draw the current screen. within
+a type of annotations, items are in sorted, ascending order, meaning
+they can be processed in order, keeping a single index into both the
+offset and payload arrays.
+
+#### Annotations are stateless
+
+To the frontend, annotations are _stateless_. By this we mean that
+a given update will include the full set of annotations that intersect
+with the currently visible region; the client will not be expected to do
+any caching or merging. (As a possible optimization we might consider
+making this be true only _per type_; so that once received, annotations
+of a given type are considered valid until new annotations of that type
+are received.)
+
+#### Each type of annotation has a defined payload type.
+
+The payload associated with a given annotation is determined by the type
+of the annotation. We anticipate defining a set of types and their
+corresponding schemas as part of the xi-editor project, and more may be
+added in the future; but new types could also be defined by a plugin,
+and as long as they're supported by a frontend that's fine.
+
+Some examples of possible types and their possible payloads:
+
+**'`selection`'**: The payload for a selection is simply the location in
+the selection of the cursor. This might be represented by a number in
+`[0, 1, 2]`, corresponding to 'start', 'end/upstream', or
+'end/downstream'.
```

Comment:
The upstream/downstream distinction is important for both start and end positions. Interestingly, it could be subsumed by using line/offset instead of byte range - "upstream" is the end of line n, while "downstream" is offset 0 of line n+1.

Reply:
Okay, I don't fully grok the upstream/downstream logic, so this might be a poor example.